aregion.cpp

```diff
@@ -2743,6 +2748,718 @@ int mapBiome(int biome) {
 	}
 }
 
+struct WaterBody {
+	int name;
+	std::unordered_set<graphs::Location2D> regions;
+	std::unordered_set<int> connections;
+	bool rivers;
+
+	bool includes(const graphs::Location2D key) {
+		return regions.find(key) != regions.end();
+	}
+
+	bool includes(ARegion* reg) {
+		return regions.find({ reg->xloc, reg->yloc }) != regions.end();
+	}
+
+	void add(const graphs::Location2D key) {
+		regions.insert(key);
+	}
+
+	void add(const ARegion* reg) {
+		regions.insert({ reg->xloc, reg->yloc });
+	}
+
+	void connect(WaterBody* other) {
+		connections.insert(other->name);
+	}
+
+	bool connected(WaterBody* other) {
+		return connections.find(other->name) != connections.end();
+	}
+};
+
+int findWaterBody(std::vector<WaterBody*>& waterBodies, ARegion* reg) {
+	for (auto water : waterBodies) {
+		if (water->includes(reg)) {
+			return water->name;
+		}
+	}
+
+	return -1;
+}
+
+bool isInnerWater(ARegion* reg) {
+	for (int i = 0; i < NDIRS; i++) {
+		auto n = reg->neighbors[i];
+		if (n && n->type != R_OCEAN) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool isNearWater(ARegion* reg) {
+	for (int i = 0; i < NDIRS; i++) {
+		auto n = reg->neighbors[i];
+		if (n && n->type == R_OCEAN) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool isNearWaterBody(ARegion* reg, WaterBody* wb) {
+	for (int i = 0; i < NDIRS; i++) {
+		auto n = reg->neighbors[i];
+		if (n && wb->includes(n)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool isNearWaterBody(ARegion* reg, std::vector<WaterBody*>& list) {
+	for (auto wb : list) {
+		if (isNearWaterBody(reg, wb)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+void makeRivers(Map* map, ARegionArray* arr, std::vector<WaterBody*>& waterBodies, std::unordered_map<ARegion*, int>& rivers,
```

Comment:
should it go to mapgen instead of region file?

Reply:
then there will be a circular dependency